sagemaker-neo-notebooks/edge/cpp-integration/tutorial.cc

```diff
@@ -0,0 +1,603 @@
+  for (int i = 0; i < attempts; i++) {
+    std::cout << "Waiting for compilation..." << std::endl;
+    auto status = poll_job_status(job_name);
+    if (status == Aws::SageMaker::Model::CompilationJobStatus::COMPLETED) {
+      std::cout << "Compile successfully" << std::endl;
+      is_success = true;
+      break;
+    } else if (status == Aws::SageMaker::Model::CompilationJobStatus::FAILED) {
+      throw std::runtime_error("Compilation fail");
+    }
+    std::this_thread::sleep_for(std::chrono::seconds(30));
+  }
+
+  if (!is_success) {
+    throw std::runtime_error("Compilation timeout");
+  }
+  std::cout << "Done!" << std::endl;
+}
+
+void GetCompiledModelFromNeo(std::string& bucket_name, std::string& model_name, std::string& target,
+                             std::string& compiled_filename) {
+  std::string output_path = "output/" + model_name + "-" + target;
+
+  Aws::String aws_bucket_name(bucket_name.c_str(), bucket_name.size());
+  Aws::String aws_object_name(output_path.c_str(), output_path.size());
+
+  // download s3 object
+  Aws::S3::S3Client s3_client = getS3Client();
+
+  Aws::S3::Model::GetObjectRequest object_request;
+  object_request.SetBucket(aws_bucket_name);
+  object_request.SetKey(aws_object_name);
+
+  auto get_object_outcome = s3_client.GetObject(object_request);
+  if (!get_object_outcome.IsSuccess()) {
+    auto error = get_object_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("GetModel error: " + error_str);
+  } else {
+    auto& model_file = get_object_outcome.GetResultWithOwnership().GetBody();
+
+    // download the sample file
+    const char* output_filename = compiled_filename.c_str();
+    std::ofstream output_file(output_filename, std::ios::binary);
+    output_file << model_file.rdbuf();
+  }
+}
+
+void DownloadNpyData() {
+  Aws::S3::S3Client s3_client = getS3Client();
+
+  // this is a preprocess image file in our sample directory
+  // this can be changed to any s3 bucket you specified
+  Aws::String aws_bucket_name = "neo-ai-dlr-test-artifacts";
+  Aws::String aws_object_name = "test-data/dog.npy";
+  std::string filename = "dog.npy";
+
+  Aws::S3::Model::GetObjectRequest object_request;
+  object_request.SetBucket(aws_bucket_name);
+  object_request.SetKey(aws_object_name);
+
+  auto get_object_outcome = s3_client.GetObject(object_request);
+  if (!get_object_outcome.IsSuccess()) {
+    auto error = get_object_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("DownloadNpyData error: " + error_str);
+  } else {
+    auto& image_file = get_object_outcome.GetResultWithOwnership().GetBody();
+    const char* output_filename = filename.c_str();
+    std::ofstream output_file(output_filename, std::ios::binary);
+    output_file << image_file.rdbuf();
+  }
+}
+
+template <typename T>
+int GetPreprocessNpyFile(std::string& npy_filename, std::vector<unsigned long>& input_shape,
```

Comment:
`const std::string& npy_filename`

are we returning data through `input_shape` and `input_data`? if not, we can make them const reference.

Reply:
yes, we are, so those stay as is